File: examples/ts/harness.ts
/**
 * Runs every lesson example and captures two things per lesson:
 *
 *   regions  - named extracts of the source, delimited by `// region: name` / `// endregion`
 *   outputs  - stdout, split into the sections the example declared with `section("name")`
 *
 * Usage (Node 24+, no build step):
 *   node harness.ts list
 *   node harness.ts run 2.2
 *   node harness.ts capture ../../web/public/data/captures/ts
 */

import { execFileSync } from 'node:child_process';
import { mkdirSync, mkdtempSync, readFileSync, readdirSync, rmSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join, relative } from 'node:path';
import { fileURLToPath } from 'node:url';
import { MARKER } from './src/learnai/index.ts';

const ROOT = fileURLToPath(new URL('.', import.meta.url));
const REPO = join(ROOT, '..', '..');
const FILE_RE = /^m(\d+)_[a-z0-9_]+\/l(\d+)_[a-z0-9_]+\.ts$/;

function discover(): Map<string, string> {
  const lessons = new Map<string, string>();
  for (const dir of readdirSync(ROOT, { withFileTypes: true })) {
    if (!dir.isDirectory() || !/^m\d+_/.test(dir.name)) continue;
    for (const file of readdirSync(join(ROOT, dir.name)).sort()) {
      const rel = `${dir.name}/${file}`;
      const match = FILE_RE.exec(rel);
      if (!match) continue;
      lessons.set(`${Number(match[1])}.${Number(match[2])}`, join(ROOT, rel));
    }
  }
  return lessons;
}

function extractRegions(source: string): Record<string, string> {
  const regions: Record<string, string> = {};
  let current: string | null = null;
  let buffer: string[] = [];
  for (const line of source.split('\n')) {
    const stripped = line.trim();
    if (stripped.startsWith('// region:')) {
      current = stripped.slice('// region:'.length).trim();
      buffer = [];
    } else if (stripped === '// endregion' && current) {
      regions[current] = dedent(buffer.join('\n'));
      current = null;
    } else if (current !== null) {
      buffer.push(line);
    }
  }
  return regions;
}

function dedent(text: string): string {
  const lines = text.split('\n');
  const indents = lines.filter((l) => l.trim()).map((l) => l.length - l.trimStart().length);
  const indent = indents.length ? Math.min(...indents) : 0;
  return lines
    .map((l) => (l.trim() ? l.slice(indent) : ''))
    .join('\n')
    .replace(/^\n+|\n+$/g, '');
}

function splitSections(stdout: string): Record<string, string> {
  const sections: Record<string, string> = {};
  let current = '_';
  let buffer: string[] = [];
  const flush = () => {
    const text = buffer.join('\n').replace(/^\n+|\n+$/g, '');
    if (current !== '_' || text) sections[current] = text;
  };
  for (const line of stdout.split('\n')) {
    if (line.startsWith(MARKER)) {
      flush();
      current = line.slice(MARKER.length).trim();
      buffer = [];
    } else {
      buffer.push(line);
    }
  }
  flush();
  return sections;
}

interface Recorded {
  model: string;
  recorded_at: string;
}

/** Run one example in its own process; return its stdout and the cassette recordings it used. */
function run(path: string): { stdout: string; recorded: Recorded[] } {
  const dir = mkdtempSync(join(tmpdir(), 'learnai-'));
  const log = join(dir, 'cassettes.jsonl');
  writeFileSync(log, '');
  try {
    const stdout = execFileSync(process.execPath, [path], {
      encoding: 'utf8',
      cwd: ROOT,
      stdio: 'pipe',
      env: { ...process.env, LEARNAI_CASSETTE_LOG: log },
    });
    const used = readFileSync(log, 'utf8').split('\n').filter(Boolean).map((l) => JSON.parse(l) as Recorded);
    const unique = new Map(used.map((u) => [`${u.model}|${u.recorded_at}`, u]));
    const recorded = [...unique.values()].sort((a, b) => `${a.model}|${a.recorded_at}`.localeCompare(`${b.model}|${b.recorded_at}`));
    return { stdout, recorded };
  } catch (error) {
    const stderr = (error as { stderr?: string }).stderr ?? '';
    console.error(`example failed: ${relative(ROOT, path)}\n${stderr}`);
    process.exit(1);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
}

function capture(outDir: string): void {
  mkdirSync(outDir, { recursive: true });
  const lessons = discover();
  for (const [lessonId, path] of lessons) {
    const source = readFileSync(path, 'utf8');
    const { stdout, recorded } = run(path);
    const payload = {
      lesson: lessonId,
      language: 'ts',
      file: relative(REPO, path).split('\\').join('/'),
      regions: extractRegions(source),
      outputs: splitSections(stdout),
      // Which model + date the recorded responses came from, if the example called one.
      recorded,
    };
    writeFileSync(join(outDir, `${lessonId}.json`), JSON.stringify(payload, null, 2));
    console.log(
      `captured ${lessonId}: ${Object.keys(payload.regions).length} regions, ${Object.keys(payload.outputs).length} outputs`,
    );
  }
  console.log(`${lessons.size} lesson(s) -> ${outDir}`);
}

const [command = 'list', arg] = process.argv.slice(2);
const lessons = discover();
if (command === 'list') {
  for (const [id, path] of lessons) console.log(`${id.padEnd(6)} ${relative(ROOT, path)}`);
} else if (command === 'run') {
  const path = arg && lessons.get(arg);
  if (!path) {
    console.error(`no example for lesson ${arg}`);
    process.exit(1);
  }
  process.stdout.write(run(path).stdout);
} else if (command === 'capture') {
  capture(arg ?? 'captures');
} else {
  console.error('usage: node harness.ts list | run <id> | capture <dir>');
  process.exit(1);
}

File: examples/python/harness.py
"""
Runs every lesson example and captures two things per lesson:

  regions  - named extracts of the source, delimited by `# region: name` / `# endregion`
  outputs  - stdout, split into the sections the example declared with `section("name")`

Usage:
  python3 -m harness list
  python3 -m harness run 2.2                       # run one lesson, print its output
  python3 -m harness capture ../../web/public/data/captures/python
  python3 -m harness cassettes                     # which lesson uses each recording; orphans
  python3 -m harness prune                         # delete recordings nothing uses
"""

from __future__ import annotations

import io
import json
import os
import re
import runpy
import sys
import tempfile
from contextlib import redirect_stdout
from pathlib import Path

from learnai import MARKER

ROOT = Path(__file__).parent
FILE_RE = re.compile(r"^m(\d+)_[a-z0-9_]+/l(\d+)_[a-z0-9_]+\.py$")


def discover() -> dict[str, Path]:
    """Map lesson id ("2.2") -> example file, from the mNN_topic/lNN_name.py convention."""
    lessons: dict[str, Path] = {}
    for path in sorted(ROOT.glob("m*_*/l*_*.py")):
        match = FILE_RE.match(path.relative_to(ROOT).as_posix())
        if not match:
            continue
        lesson_id = f"{int(match.group(1))}.{int(match.group(2))}"
        lessons[lesson_id] = path
    return lessons


def extract_regions(source: str) -> dict[str, str]:
    regions: dict[str, str] = {}
    current: str | None = None
    buffer: list[str] = []
    for line in source.splitlines():
        stripped = line.strip()
        if stripped.startswith("# region:"):
            current = stripped.removeprefix("# region:").strip()
            buffer = []
        elif stripped == "# endregion" and current:
            regions[current] = dedent("\n".join(buffer))
            current = None
        elif current is not None:
            buffer.append(line)
    return regions


def dedent(text: str) -> str:
    lines = [l for l in text.splitlines() if l.strip()]
    if not lines:
        return text
    indent = min(len(l) - len(l.lstrip()) for l in lines)
    return "\n".join(l[indent:] if l.strip() else "" for l in text.splitlines()).strip("\n")


def split_sections(stdout: str) -> dict[str, str]:
    sections: dict[str, str] = {}
    current = "_"
    buffer: list[str] = []
    for line in stdout.splitlines():
        if line.startswith(MARKER):
            sections[current] = "\n".join(buffer).strip("\n")
            current = line.removeprefix(MARKER).strip()
            buffer = []
        else:
            buffer.append(line)
    sections[current] = "\n".join(buffer).strip("\n")
    sections.pop("_", None) if not sections.get("_") else None
    return sections


def run(path: Path) -> tuple[str, list[dict]]:
    """Run one example; return its stdout and the list of cassette recordings it used."""
    out = io.StringIO()
    with tempfile.NamedTemporaryFile("w+", suffix=".jsonl", delete=False) as log:
        os.environ["LEARNAI_CASSETTE_LOG"] = log.name
        try:
            with redirect_stdout(out):
                runpy.run_path(str(path), run_name="__main__")
        finally:
            os.environ.pop("LEARNAI_CASSETTE_LOG", None)
        log.seek(0)
        used = [json.loads(line) for line in log if line.strip()]
    os.unlink(log.name)
    return out.getvalue(), used


def capture(out_dir: Path) -> None:
    out_dir.mkdir(parents=True, exist_ok=True)
    lessons = discover()
    for lesson_id, path in lessons.items():
        source = path.read_text(encoding="utf-8")
        stdout, used = run(path)
        payload = {
            "lesson": lesson_id,
            "language": "python",
            "file": path.relative_to(ROOT.parent.parent).as_posix(),
            "regions": extract_regions(source),
            "outputs": split_sections(stdout),
            # Which model + date the recorded responses came from, if the example called one.
            "recorded": [{"model": m, "recorded_at": d} for m, d in sorted({(u["model"], u["recorded_at"]) for u in used})],
        }
        (out_dir / f"{lesson_id}.json").write_text(json.dumps(payload, indent=2), encoding="utf-8")
        print(f"captured {lesson_id}: {len(payload['regions'])} regions, {len(payload['outputs'])} outputs")
    print(f"{len(lessons)} lesson(s) -> {out_dir}")


def cassettes(prune: bool = False) -> int:
    """Audit examples/shared/cassettes: which lesson uses each recording, and which recordings
    nothing uses any more (safe to delete). `prune` deletes the orphans."""
    from learnai.llm import CASSETTES  # noqa: PLC0415

    used_by: dict[str, set[str]] = {}
    targets = {**discover(), "smoke": ROOT / "smoke_llm.py"}  # the adapter smoke test has a fixture too
    for lesson_id, path in targets.items():
        try:
            _, used = run(path)
        except Exception as e:  # noqa: BLE001 - an unrecorded lesson is reported, not fatal
            print(f"{lesson_id:6} could not run: {str(e).splitlines()[0][:80]}")
            continue
        for u in used:
            used_by.setdefault(u["key"], set()).add(lesson_id)
    on_disk = sorted(p.stem for p in CASSETTES.glob("*.json"))
    orphans = [k for k in on_disk if k not in used_by]
    for key in on_disk:
        meta = json.loads((CASSETTES / f"{key}.json").read_text(encoding="utf-8"))
        who = ", ".join(sorted(used_by.get(key, ()))) or "UNUSED"
        print(f"{key}  {meta['response']['model']:12} {meta['recorded_at']}  {who}")
    print(f"{len(on_disk)} cassettes, {len(used_by)} in use, {len(orphans)} unused")
    if prune:
        for key in orphans:
            (CASSETTES / f"{key}.json").unlink()
        print(f"deleted {len(orphans)} unused cassette(s)")
    return 0


def main(argv: list[str]) -> int:
    command = argv[1] if len(argv) > 1 else "list"
    lessons = discover()
    if command == "list":
        for lesson_id, path in lessons.items():
            print(f"{lesson_id:6} {path.relative_to(ROOT)}")
        return 0
    if command == "run":
        lesson_id = argv[2]
        if lesson_id not in lessons:
            print(f"no example for lesson {lesson_id}", file=sys.stderr)
            return 1
        sys.stdout.write(run(lessons[lesson_id])[0])
        return 0
    if command == "capture":
        capture(Path(argv[2] if len(argv) > 2 else "captures"))
        return 0
    if command in ("cassettes", "prune"):
        return cassettes(prune=command == "prune")
    print(__doc__)
    return 1


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
